walnuts_cpp/nuts.hpp

```diff
@@ -30,55 +26,49 @@ class Span {
   Vec<S> theta_select_;
   S logp_;
 
-  // assumes _bk and _fw ae distinct; see other ctor
-  Span(const Vec<S> theta_bk,
-       const Vec<S> rho_bk,
-       const Vec<S> grad_theta_bk,
-       const Vec<S> theta_fw,
-       const Vec<S> rho_fw,
-       const Vec<S> grad_theta_fw,
-       const Vec<S> theta_select,
-       S logp)
-      : theta_bk_(std::move(theta_bk)),
-        rho_bk_(std::move(rho_bk)),
-        grad_theta_bk_(std::move(grad_theta_bk)),
-        theta_fw_(std::move(theta_fw)),
-        rho_fw_(std::move(rho_fw)),
-        grad_theta_fw_(std::move(grad_theta_fw)),
-        theta_select_(std::move(theta_select)),
-        logp_(logp)
-  {}
-
-  Span(const Vec<S> theta,
-       const Vec<S> rho,
-       const Vec<S> grad_theta,
+  Span(const Vec<S>& theta,
+       Vec<S>& rho,
+       Vec<S>& grad_theta,
        S logp)
       : theta_bk_(theta),  // copy duplicates
         rho_bk_(rho),
         grad_theta_bk_(grad_theta),
         theta_fw_(theta),
         rho_fw_(std::move(rho)),  // move once after copy
         grad_theta_fw_(std::move(grad_theta)),
-        theta_select_(std::move(theta)),
+        theta_select_(theta),
         logp_(logp)
   {}
 
-  Span(const Span<S> span1,
-       const Span<S> span2,
-       const Vec<S> theta_select,
+  Span(Vec<S>& theta,
+    Vec<S>& rho,
+    Vec<S>& grad_theta,
+    S logp)
+   : theta_bk_(theta),  // copy duplicates
+     rho_bk_(rho),
+     grad_theta_bk_(grad_theta),
+     theta_fw_(theta),
+     rho_fw_(std::move(rho)),  // move once after copy
+     grad_theta_fw_(std::move(grad_theta)),
```

Comment:
strictly this might cause issues as rho is passed by reference

Reply:
It’s fine as it’s actually used in the code, but we could make it clearer by requiring &&s (not that C++ really checks that you’re handling those correctly, anyway…)

Reply:
Thanks for jumping in @DiamonDinoia.  I was hoping to wait until we had a pretty well optimized set up before starting to bug you about it!